### ISSUE TITLE
Simplify getting output labels

### ISSUE THREAD

**Issue by fkodom** [IDEA-Research/GroundingDINO#11]
First, just wanted to say thanks for publishing this repo -- very cool and love that it's fully open-sourced.

I wanted to suggest a simplification to `get_phrases_from_posmap`.  For reference, here is the  [current definition](https://github.com/IDEA-Research/GroundingDINO/blob/858efccbad1aed50644f0185e49f4254a9af7560/groundingdino/util/utils.py#L598):
```python
def get_phrases_from_posmap(posmap: torch.BoolTensor, tokenlized, caption: str):
    assert isinstance(posmap, torch.Tensor), "posmap must be torch.Tensor"
    if posmap.dim() == 1:
        non_zero_idx = posmap.nonzero(as_tuple=True)[0].tolist()
        words_list = caption.split()

        # build word idx list
        words_idx_used_list = []
        for idx in non_zero_idx:
            word_idx = tokenlized.token_to_word(idx)
            if word_idx is not None:
                words_idx_used_list.append(word_idx)
        words_idx_used_list = set(words_idx_used_list)

        # build phrase
        words_used_list = []
        for idx, word in enumerate(words_list):
            if idx in words_idx_used_list:
                words_used_list.append(word)

        sentence_res = " ".join(words_used_list)
        return sentence_res
    else:
        raise NotImplementedError("posmap must be 1-dim")
```

It seems like this is over-complicating the text decoding, since we can already do that with the tokenizer.  An interesting edge case with your implementation:
- If the input label contains a long word (e.g. "American flag"), the long word is often lost in the output label.  I believe this is because the long word ("American") is split into multiple tokens.  In that case, the `for idx, word in enumerate(words_list)` loop is incorrect, because the positional index of the word is offset.  In my tests, it has often returned incorrect or empty text labels, due to this issue.

A potential fix:

```python
from typing import Dict

import torch
from transformers import AutoTokenizer

def get_phrases_from_posmap(
    posmap: torch.BoolTensor, tokenized: Dict, tokenizer: AutoTokenizer
):
    assert isinstance(posmap, torch.Tensor), "posmap must be torch.Tensor"
    if posmap.dim() == 1:
        non_zero_idx = posmap.nonzero(as_tuple=True)[0].tolist()
        token_ids = [tokenized["input_ids"][i] for i in non_zero_idx]
        return tokenizer.decode(token_ids)
    else:
        raise NotImplementedError("posmap must be 1-dim")
```

We don't have to compare against the original caption, because all token information is already contained in the tokenizer.  This gives the correct result "American flag" in my example above.

**Comment by paulguerrie** [comment]
@fkodom Great fix!  I was just about to dive into that function myself when I saw this issue.  Wanted to post a before/after example to demonstrate the difference.

Here's a snippet of my code (pulled from the demo Colab, modified):
```
TEXT_PROMPT = "chair, dog, shoe, hat, car, light bulb, phone, coffee cup, table, street sign"
BOX_TRESHOLD = 0.35
TEXT_TRESHOLD = 0.25

image_source, image = load_image(IMAGE_PATH)

boxes, logits, phrases = predict(
    model=model, 
    image=image, 
    caption=TEXT_PROMPT, 
    box_threshold=BOX_TRESHOLD, 
    text_threshold=TEXT_TRESHOLD
)
```

Here's before applying the proposed fix to `get_phrases_from_posmap`:
![before](https://user-images.githubusercontent.com/97041392/228059927-0a6ccccb-e24c-4e84-a543-cee491f5bb8e.png)

Here's after applying the fix:
![after](https://user-images.githubusercontent.com/97041392/228059972-d8a46203-6f84-4a2a-90b6-b14ca027da19.png)

**Comment by fkodom** [comment]
🙌

**Comment by SkalskiP** [comment]
I'll create a PR, and wait for @SlongLiu to evaluate it.

**Comment by SkalskiP** [comment]
In that case, I'll go ahead and close the issue. @paulguerrie and @fkodom if you'll continue to experience problems like this feel free to reopen it.
